### ISSUE TITLE
Box-DDP - [merged]

### ISSUE THREAD

**Issue by wxmerkt** [loco-3d/crocoddyl#575]
In GitLab by @wxmerkt on Oct 3, 2019, 15:39

_Merges topic/wxm-boxddp -> devel_

Work-in-progress for Box-DDP. It works for some examples (quadruped walking), but not for others (jumping etc.).

Quadruped walking within control limits:

![20191003_boxddp_walking_within_control_limits](/uploads/faeff1d0660777a4a4a98875fb0c5adc/20191003_boxddp_walking_within_control_limits.png)

Current issues:

1. The Box-QP is a quick implementation and does allocate memory during runtime.
2. The stopping criteria don't work right now. Walking is within limits within ~400 iterations, but it will happily continue optimising for many, many more.

Cost convergence plot for quadruped walking (log scale):

![20191003_boxddp_walking_cost_log_scale](/uploads/95c83c6dc048959e7d5e38cb82dfecf7/20191003_boxddp_walking_cost_log_scale.png)

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:44

You should use the norm Inf here (independent from the length of the trajectory)

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:45

Pre-allocate here, as you know the matrix dimension to avoid branching latter.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:45

You should use noalias()

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:47

You should use a tmp variable declare outside of the loop before doing the inversion.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:48

For the inversion, prefer Cholesky decomposition instead of raw inverse

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:48

You can use noalias() here again

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:48

You should declare as a static member.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:49

Use temporary here to store intermediate result

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:49

Set the size of the vector here

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:50

This seems to be a copy past of the internal code of Edinburgh. It would be worth to use the same template as crocodyl.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:50

Crocodyle namespace here

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:52

pow is very slow. Prefer to use the standard way of obtaining the power of 2.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 15:52

noalias

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:09

Let's do use `Eigen::Ref` since we cannot properly binding with EigenPy. Do it like this
```cpp

  void set_u_lower_limit(const Eigen::VectorXd& u_in);
  void set_u_upper_limit(const Eigen::VectorXd& u_in);
```

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 16:10

We can use Eigen::Ref for vectors, not for Matrix :(

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:11

same here

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:11

Please add CNRS-LAAS too :)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:12

remove override keywords

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:13

follow convention of guard names.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:14

It would be great if you create a data structure and allocated together with the `SolverBoxDDP`. And you should allocate variables to avoid them to be create at runtime. I see @jcarpent is pointing out some of these stuffs.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:17

If I am right, this is not c++98. Please remove constexpr

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 16:18

Yes, this is a quick test whether it works (and far from a strong or numerically stable implementation). Thank you very much for all your feedback and review.
Carlos asked for the code to be pushed up as soon as possible. It is definitely not ready for merging yet.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:19

This is computed only once, not important but why not!

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 16:20

Yes, as mentioned over email. This is a test - it works but far from ready to be merged. We pushed it here as a pull request to have a basis for discussion. We are aware that this does have lots of temporary allocations

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:22

use brackets here. In general, you need to format the code.

See instructions here: https://gepgitlab.laas.fr/loco-3d/crocoddyl/wikis/Developer-Guide

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:24

Why is there not any regularization scheme?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:29

let's call it: "u_lb" and "u_ub". I used name for complete objects such as state.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:30

As mentioned here
https://gepgitlab.laas.fr/loco-3d/crocoddyl/merge_requests/236#note_6869

Do the same for functions,e .g.
```cpp
 const Eigen::VectorXd& get_u_ub() const;
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:31

Ok, I wasn't updated. @jcarpent -- Is this feature released in robotpkg?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:32

same as https://gepgitlab.laas.fr/loco-3d/crocoddyl/merge_requests/236#note_6869

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:33

We don't need to copy the entire license.

Copy this text
```cpp
///////////////////////////////////////////////////////////////////////////////
// BSD 3-Clause License
//
// Copyright (C) 2018-2019, LAAS-CNRS, The University of Edinburgh
// Copyright note valid unless otherwise stated in individual files.
// All rights reserved.
///////////////////////////////////////////////////////////////////////////////
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 16:41

### Jumping motions and DDP
It's expected that jumping motion doesn't work. This solver is based on DDP which poorly finds solutions for these motions. However, I believe we have to do it later. For that, could you add the Python scripts that you use?

### Jumping motions and impulse dynamics
In practice, we've found a greater convergence properties when we use impulse dynamics; including with the DDP solver.

**Note**

The impulse dynamics doesn't have damping term, which makes them not useful now. Adding them is in my priority list, and the only limitations is robotpkg.


### BoxDDP limitations
We already knew that there is a problem in convergent of BoxDDP. We believe that our expected model is not good enough. I'm working on a new expected model, having your scripts I could validate it.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_59_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_55_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_65_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_107_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_107_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_110_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_115_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_114_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_118_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_161_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_159_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_20_20)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_58_57)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#a967a34680cc6af30573872d7c6ab72f96f8835f_47_47)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#9a7b7cf46b59c6d8aae969ae8b120b785542cebb_63_63)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#25a64810559db631805e9b3258de9bd1e63607fa_4_4)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#25a64810559db631805e9b3258de9bd1e63607fa_26_26)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_30_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_46_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_153_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_77_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#b1933866d010f2155cccd3271fe83b8f063cfecf_105_105)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#9a7b7cf46b59c6d8aae969ae8b120b785542cebb_59_59)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#a967a34680cc6af30573872d7c6ab72f96f8835f_55_51)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

changed this line in [version 2 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b#4379a156b071541ec1e710b16c1686af83d0b979_1_0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

added 4 commits

<ul><li>b3dd26d4 - Change u_lower_limit to u_lb, u_upper_limit to u_ub</li><li>09dd103e - DDP: Avoid pow</li><li>20d2efe4 - BoxDDP: Updates from review</li><li>8211a907 - BoxQP: Updates from review, rename, format</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1197&start_sha=66b005be2d3a13a9943c7d5ed098e4de6ba1c84b)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:28

We based this on Yuval Tassa's BoxQP and he used Euclidean norm. I've changed to infinity norm now.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:29

Hff, Hfc, Hff_inv change dimension in the current implementation (can and should be changed)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 17:58

added 1 commit

<ul><li>7e805607 - Add quadruped BoxDDP example script</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1198&start_sha=8211a9070d216298f30353386a3db6a95e3ce895)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 18:18

added 1 commit

<ul><li>b3a98a69 - DDP/BoxDDP: Revert pow change</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1199&start_sha=7e80560733053ec43f27990700673b4606f156e6)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 18:21

Do you refer to adaptation of lambda? We implemented the most basic variant here. I'm planning to add interfaces to proper solvers next

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 18:27

Yes, I do. 

That would be great. I believe the `SolverBoxDDP` should pass this value. In Crocoddyl, we have two kind of regularizations: control and state. For this case, we're talking about the control one.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 18:29

I do pass `ureg_` from `SolverBoxDDP` as `lambda`. However, it is constant during the BoxQP iterations and does not get changed.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 3, 2019, 18:29

@cmastalli here we pass `ureg_` as the final argument as the value for `lambda`

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 18:49

Good!

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 3, 2019, 18:49

Thanks for pointing out the code, I really appreciate this!

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 3, 2019, 20:18

To avoid memory re-allocation, you then need to allocate with the max size you know. 
Then, you should use a view o topLeftCorner view from Eigen::Matrix.

It will make your code without memory allocation then.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 06:40

Better to use `unsigned int`

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 06:40

same here

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:04

Here you have the problem. An integrated action is defined as an action model, not differential action model. So the dimension of u_am and u_dam aren't strictly the same, e.g. floating-base systems.

I would need to improve our diff-action API to handle this properly. For this PR, please add an if-declaration, e.g.
```cpp
      differential_(model),
      time_step_(time_step),
      time_step2_(time_step * time_step),
      with_cost_residual_(with_cost_residual) {
  // TODO(cmastalli): handling properly at the diff-action API level
  unsigned int const& ntau = differential_->get_nu();
  if (ntau < nu_) {
    Eigen::VectorXd lb = Eigen::VectorXd::Constant(nu_,-std::numeric_limits<double>::infinity());
    lb.tail(ntau) = differential_->get_u_lb();
    set_u_lb(lb);
  } else if (ntau > 0) // see note
    set_u_lb(differential_->get_u_lb());
  }
  // Note: ntau could be zero, e.g. impulse dynamics. Crocoddyl supports OC problems with a mixture of autonomous and non-autonomous dynamics

  // Same for the upper bound
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:09

```cpp
unsigned int const&
```

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:09

same as https://gepgitlab.laas.fr/loco-3d/crocoddyl/merge_requests/236#note_6915

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:09

same as https://gepgitlab.laas.fr/loco-3d/crocoddyl/merge_requests/236#note_6915

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:13

Allocate `low_limit` and `high_limit` only once, and inside `allocateData`. Rename these variables as `u_ll_` and `u_hl_`.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:15

You don't need noalias() in this line.

We use it in multiplication operations to avoid creation of temporal objects done by Eigen.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 07:26

It would be nice if
 1. you includes the inputs that we define to the other scripts.
 2. add the example in the cmake file. 

See the README, it explains how to run example with display and plot option.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 4, 2019, 07:56

This is not a real issue here. It won't change anything in terms of performances to put or not the noalias()

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 11:24

What is our reason for preferring `unsigned int`? To get underflow protection or for modulo math?

From Bjarne Stroustrup (creator of C++): "Use [signed] int until you have a reason not to. Use unsigned if you are fiddling with bit patterns, and never mix signed and unsigned."

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 11:25

I take it, it is safe now to use `Eigen::Ref<const Eigen::VectorXd>&`?

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

changed this line in [version 5 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_49_54)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

changed this line in [version 5 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec#319d2fb71e3a722bd2b38fb40123a890ec160675_20_20)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

changed this line in [version 5 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_29_29)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

changed this line in [version 5 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_33_34)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

changed this line in [version 5 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec#cd4dfd96267fcfb7fa3f7a69756ea172929fdddf_47_53)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 12:36

added 5 commits

<ul><li>bfeb3dc1 - Update quadrupedal walk example with BoxDDP</li><li>49c51c10 - Update copyrights</li><li>03137d46 - BoxDDP: Address code review</li><li>298568b2 - IntegratedActionModelEuler: Proper assignment of control limits</li><li>30517548 - DifferentialActionModelAbstract: Fix allocation bug</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1201&start_sha=b3a98a69e5b6ce3375cb9108def0503354be1eec)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 12:44

To avoid that users pass negative values.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 12:46

You can double check it yourself. Write a script that set and get the lower bounds, if there is a problem it wouldn't return the vector.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 15:22

added 1 commit

<ul><li>f85ae833 - Fix flake...</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1207&start_sha=30517548f2bb31f3928cf8a1037b966551d74155)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 15:24

added 8 commits

<ul><li>f85ae833...15cd5de7 - 7 commits from branch <code>loco-3d:devel</code></li><li>f86cf25f - Merge remote-tracking branch &#39;origin/devel&#39; into topic/wxm-boxddp</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1208&start_sha=f85ae833ee8f796a0b48876fad26de9951ae7519)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 15:47

I don't think that is a good way of doing it. Doesn't `-1` for an `unsigned` represent `UINT_MAX`?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 16:22

No, but we could do it later.

I'm not against of -1 equals to unbounded iterations.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:24

I think we should use a `if` statement to check and leave it as an `int`. It is fine with `unsigned` for Python (gives me an `Overflow` error for `-1`), while in C++ it would underflow to `UINT_MAX`.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:27

@jcarpent @cmastalli I have rewritten the BoxQP to separate data (and allocate it before, pass the data struct) and then use this data struct in the algorithm (passed as a reference). However, this slows the BoxQP _down_ considerably. As it is right now (with `inline BoxQP`), it takes ~7.6microseconds while with the separation it takes ~38microseconds. Is it possible that we have cache misses with the separation?

Due to it performing worse, I am not sure whether to push it.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 16:27

sounds good. I will later propagate to the other solvers.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 16:32

Difficult to say, it would be important to check if the algorithm has cache misses.

You can always push these stuffs in another branch, where we can check it.

**Comment by wxmerkt** [comment]
In GitLab by @jcarpent on Oct 4, 2019, 16:42

It is indeed weird. Why do you inline it? This may slow the performances as your code is very long.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:42

added 1 commit

<ul><li>13152b2f - BoxDDP: Fix flake</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1209&start_sha=f86cf25f54427db4d6c05e2c2bf28c4436a6b9c0)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:43

Sorry, the other way around. I inline the current version but not the separated data. Let me open a new PR for you to have a look at

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:45

@jcarpent @cmastalli Cf. here: https://gepgitlab.laas.fr/memmo/crocoddyl/merge_requests/1 . This is the diff to this PR.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:50

added 1 commit

<ul><li>a09b43d5 - BoxQP: More review updates</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1211&start_sha=13152b2f4185c3438c0642b18570da2215cb5c16)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 4, 2019, 16:51

Thank you very much for all the feedback! I am done with the updates to the review comments.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 4, 2019, 19:30

assigned to @cmastalli

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 11:35

You might delete this by mistake during solving conflicts, did you? 

Could you put it back?

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 11:40

What is the status of this task?

We can also work on it in another PR.

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 11:41

@wxmerkt thank you for this great effort. We're almost there as you can see.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:23

changed this line in [version 10 of the diff](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1216&start_sha=a09b43d5570f9f0b750c655225323d754295c56a#e7de2e2e099503c28f0b18ed87b13242fea9f600_56_56)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:23

added 1 commit

<ul><li>3693ecd1 - Bindings: Readd accidentally deleted ActivationSmoothAbs</li></ul>

[Compare with previous version](/loco-3d/crocoddyl/merge_requests/236/diffs?diff_id=1216&start_sha=a09b43d5570f9f0b750c655225323d754295c56a)

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:23

Yes, sorry this got lost when merging the master back into the branch and resolving conflicts. I've added it back.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:24

Given that the data separation made it _worse_ (as in the above linked PR), I am proposing to leave it as is and merge this PR. We can then look into alternate QP solvers or optimising the Box-QP in a follow-up.

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:24

resolved all threads

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:24

unmarked as a **Work In Progress**

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:24

changed the description

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 15:24

I'd really appreciate if we can get this merged early tomorrow morning. We have a few items we'd like to continue to build on top of this :-)

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 16:25

enabled an automatic merge when the pipeline for 3693ecd1fd269b52e7c7c2f6047856cdf15e309f succeeds

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 16:46

merged

**Comment by wxmerkt** [comment]
In GitLab by @cmastalli on Oct 6, 2019, 16:46

mentioned in commit 897ded9ad4b5ebc1151fde0d30aaa9118f7c5ab3

**Comment by wxmerkt** [comment]
In GitLab by @wxmerkt on Oct 6, 2019, 16:46

Awesome, thank you Carlos!
